src/reflect/scala/reflect/internal/Names.scala

```diff
@@ -515,28 +523,9 @@ trait Names extends api.Names {
   //   final override def isOperatorName = false
   // }
 
-  /** TermName_S and TypeName_S have fields containing the string version of the name.
-   *  TermName_R and TypeName_R recreate it each time toString is called.
-   */
-  private final class TermName_S(index0: Int, len0: Int, next0: TermName, override val toString: String) extends TermName(index0, len0, next0) {
-    protected def createCompanionName(next: TypeName): TypeName = new TypeName_S(index, len, next, toString)
-    override def newName(str: String): TermName = newTermNameCached(str)
-  }
-  private final class TypeName_S(index0: Int, len0: Int, next0: TypeName, override val toString: String) extends TypeName(index0, len0, next0) {
-    override def newName(str: String): TypeName = newTypeNameCached(str)
-  }
-
-  private final class TermName_R(index0: Int, len0: Int, next0: TermName) extends TermName(index0, len0, next0) {
-    protected def createCompanionName(next: TypeName): TypeName = new TypeName_R(index, len, next)
-    override def toString = new String(chrs, index, len)
-  }
-
-  private final class TypeName_R(index0: Int, len0: Int, next0: TypeName) extends TypeName(index0, len0, next0) {
-    override def toString = new String(chrs, index, len)
-  }
 
   // SYNCNOTE: caller to constructor must synchronize if `synchronizeNames` is enabled
-  sealed abstract class TermName(index0: Int, len0: Int, val next: TermName) extends Name(index0, len0) with TermNameApi {
+  final class TermName(index0: Int, len0: Int, val next: TermName, cachedString: String) extends Name(index0, len0, cachedString) with TermNameApi {
```

Comment:
Would there be a benefit in reducing either `index0` or `len0` to `Short` or `Byte`? I do not think there are any identifiers over 127 characters long.

Reply:
index0 cant be. Its an index into a blob of all of the identifiers. I might be more inclined to consider if 2G is big enough to cope with all of the names that a project could reference (It seems a strange and arbitrary limit to build in to the architecture)

What is the language limit, or the JVM limits? I don't think that we could impose a smaller limit for the length of the identifier, after all we have to be able to read a java identifier, and cope with any legal identifier in scala, and I think that is 65536, from below

from https://docs.oracle.com/javase/specs/jvms/se12/html/jvms-4.html#jvms-4.11
> The length of field and method names, field and method descriptors, and other constant string values (including those referenced by ConstantValue (§4.7.2) attributes) is limited to 65535 characters by the 16-bit unsigned length item of the CONSTANT_Utf8_info structure (§4.4.7).

Reply:
Perhaps you could use 24 bits for the `index0` and 8 bits for the `len0`, or any such arrangement, so you can fit them inside a single `Int` field. That would allow you to have  16777216 characters, for words of up to 255 characters each.

Reply:
That sounds dangerous simply because while we _could_ forbid 256-character scala names, we can't impose such a restriction on Java (and being unable even to parse such a classfile does not sound acceptable to me). We could have an `XL(Term|Type)Name` to accommodate that, perhaps.